adr/0026-schema.md

```diff
@@ -0,0 +1,569 @@
+
+- `setVariable` deprecated in favor of `setVariables`
+- `scripts` deprecated in favor of `actions`
+- `required` soon to be deprecated in favor of `optional`.
+- `group` deprecated in favor of `flavor`, however these work functionally different and cannot be automatically migrated
+- `cosignKeyPath` deprecated in favor of specifying the key path at create time
+
+Zarf has not disabled any deprecated keys thus far. On create the user is always warned when using a deprecated field, however the field still exists in the schema and functions properly. Some of the deprecated keys can be migrated automatically as the old item is a subset or directly related to it's replacement. For example, setVariable is automatically migrated to a single item list of setVariables. This migration occurs on the zarf.yaml fed into the package during package create, however the original field is not deleted from the packaged zarf.yaml because the Zarf binary used in the airgap or delivery environment is not assumed to have the new schema fields that the deprecated key was migrated to.
+
+There are two problems this ADR aims to solve:
+- Supporting deprecated features forever is complex and time consuming
+- Dropping feature support can frustrate or confuse users
+
+## Decision
+
+Zarf will begin having proper schema versions. A top level key, `apiVersion`, will be introduced to allow users to specify the schema. At the release of v1 the only valid user input for `apiVersion` will be v1. Zarf will not allow users to build at the pre v1 version. `zarf package create` will fail if the user has deprecated keys or if `apiVersion` is missing and the user will be instructed to run the new `zarf dev update-schema` command. `zarf dev update-schema` will automatically migrate deprecated fields in the users `zarf.yaml` where possible. It will also add the apiVersion key and set it to v1.
+
+The existing go types which comprise the Zarf schema will be moved to types/alpha and will never change. An updated copy of these types without the deprecated fields will be created in a package types/v1 and all schema changes will happen on these objects. Any place in the code that is using a struct included in the Zarf schema will change from `types.structName` to `v1.structName`.
+
+All deprecated features will cause an error on create. Deprecated features with a direct migration path will still be deployed if the package was created v1, as migrations will add the non deprecated fields. If a feature does not have a direct automatic migration path (cosignKeyPath & groups) the package will fail on deploy. This will happen until the alpha schema is entirely removed from Zarf, which will happen one year after v1 is released.
+
+Any key that exists at the introduction of v1 will last the entirety of that schema lifetime. The features may be deprecated, but will not be removed until the next schema version.
```

Comment:
I would love to see some expansion on this and maybe a heading for deprecation process going forward, it seems like you're alluding that it will always take two schema versions (major, minor?) to deprecate a user facing feature

Reply:
I was thinking it would just be one major version. So that contract would be anything in V1 at the start of v1 will be there until v2. Even if it's deprecated it will still be "supported" until v2 where it would be removed